File: src/main/java/arcana/common/aspects/graph/nodes/ItemNode.java
package arcana.common.aspects.graph.nodes;

import arcana.common.aspects.AspectList;
import net.minecraft.item.Item;

import java.util.ArrayList;
import java.util.List;

public class ItemNode {
    public final Item item;
    public AspectListCached cache;
    public List<IngredientNode> links = new ArrayList<>();

    public AspectListCached cacheCrafts;
    public ItemNode remainder = null;  //"bucket" and "milk" are "remainder" and "origin" respectively. One item can have several origins but only one remainder.
    public List<ItemNode> origins = new ArrayList<>();

    public ItemNode(Item item, AspectList list) {
        this.item = item;
        cache = list != null ? new AspectListCached(list) : new AspectListCached();
        cacheCrafts = item.hasCraftingRemainingItem() ? new AspectListCached() : cache;
    }

    public void calcBasic(AspectList list) {
        cache.update(list);
    }

    public void calcCraftsValue() {
        cacheCrafts.update(cache.list.copy().subtract(remainder.cache.list));
    }

    public boolean knownBasic() {
        return cache.value != Integer.MAX_VALUE;
    }

    public boolean knownCrafts() {
        return cacheCrafts.value != Integer.MAX_VALUE;
    }

    @Override
    public String toString() {
        return "Item{" + item.getRegistryName() + '}';
    }
}

File: src/main/kotlin/arcana/common/aspects/graph/nodes/ItemNode.kt
package arcana.common.aspects.graph.nodes

import arcana.common.aspects.AspectList
import net.minecraft.item.Item

class ItemNode(val item: Item, list: AspectList?) {
    var cache: AspectListCached
    var links: MutableList<IngredientNode> = ArrayList()
    var cacheCrafts: AspectListCached
    var remainder: ItemNode? =
        null //"bucket" and "milk" are "remainder" and "origin" respectively. One item can have several origins but only one remainder.
    var origins: MutableList<ItemNode> = ArrayList()

    init {
        cache = list?.let { AspectListCached(it) } ?: AspectListCached()
        cacheCrafts = if (item.hasCraftingRemainingItem()) AspectListCached() else cache
    }

    fun calcBasic(list: AspectList) {
        cache.update(list)
    }

    fun calcCraftsValue() {
        cacheCrafts.update(cache.list.copy().subtract(remainder!!.cache.list))
    }

    fun knownBasic(): Boolean {
        return cache.value != Int.MAX_VALUE
    }

    fun knownCrafts(): Boolean {
        return cacheCrafts.value != Int.MAX_VALUE
    }

    override fun toString(): String {
        return "Item{" + item.registryName + '}'
    }
}